updated MiniO-upload helper function and added filehash function :

iamanishx opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**
#3729 

Fixes #<!--Add related issue number here.-->

**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**
updated helper function according to the api requirements : 

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

## Checklist

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass


**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- The upload process now indicates whether an upload is required.
	- File integrity is enhanced with an integrated hash calculation.
- **Refactor**
	- Streamlined file handling by using a unique identifier for uploaded items.
	- Improved image selection during group chat creation.
- **Tests**
	- Comprehensive unit tests for the new hash calculation function.
	- Updated scenarios to align with the revised upload flow and response validations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

iamanishx commented:
@adithyanotfound you have to update the `createGroupChat.tsx` and the `spec file` cause mutation has been updated 
please contanct on slack or comment on docs if you need any assist

iamanishx commented on src/components/UserPortal/CreateGroupChat/CreateGroupChat.tsx:
```diff
@@ -158,8 +158,8 @@ export default function CreateGroupChat({
       try {
-        const { fileUrl } = await uploadFileToMinio(file, currentOrg);
-        setSelectedImage(fileUrl);
+        const { objectName } = await uploadFileToMinio(file, currentOrg);
+        setSelectedImage(objectName);
```
is this a better approach ?

iamanishx commented:
@palisadoes please review the test will fail as the mutation has been updated @adithyanotfound will fix as it was in part of his issue .

adithyanotfound commented:
@iamanishx Apologies! I should have marked my PR as a draft. I'll revert it once a reviewer confirms it's okay to do so.

iamanishx commented:
> @iamanishx Apologies! I should have marked my PR as a draft. I'll revert it once a reviewer confirms it's okay to do so.

No issues you can make the changes after the confirmation thank you !!

iamanishx commented:
@palisadoes  please this PR needs to merged first cause the mutation has been updated here and other will need to upgrade their resolver accordingly.

iamanishx commented on src/utils/MinioUpload.ts:
```diff
@@ -1,50 +1,58 @@
-      method: 'PUT',
-      body: file,
-      headers: {
-        'Content-Type': file.type,
-      },
-    });
+      if (!data?.createPresignedUrl) {
+        throw new Error('Failed to get presigned URL');
+      }
 
-    if (!response.ok) {
-      throw new Error('File upload failed');
-    }
+      const { presignedUrl, objectName, requiresUpload } =
+        data.createPresignedUrl;
 
-    return { fileUrl, objectName };
+      if (requiresUpload && presignedUrl) {
+        const response = await fetch(presignedUrl, {
+          method: 'PUT',
+          body: file,
+          headers: {
+            'Content-Type': file.type,
+          },
+        });
+
+        if (!response.ok) {
+          throw new Error('File upload failed');
+        }
+      }
+      return { objectName, fileHash };
+    } catch (error) {
+      console.error('Error in file upload process:', error);
+      throw error;
+    }
   };
 
   return { uploadFileToMinio };
-};
+};
```
the eslint is not working in my local machine 
it gives error like this 

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'eslint-plugin-vitest' imported from C:\Users\manis\talawa-admin\eslint.config.js
    at packageResolve (node:internal/modules/esm/resolve:854:9)
    at moduleResolve (node:internal/modules/esm/resolve:927:18)
    at defaultResolve (node:internal/modules/esm/resolve:1157:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:383:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:352:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:227:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:87:39)
    at link (node:internal/modules/esm/module_job:86:36)
PS C:\Users\manis\talawa-admin>

varshith257 commented:
Can you add few tests to validate calculatehash function?

iamanishx commented:
@coderabbitai how can i solve this issue ?
PS C:\Users\manis\talawa-admin> npx eslint --fix        

C:\Users\manis\talawa-admin\vitest.setup.ts
   5:1   error  'afterEach' is not defined  no-undef
  10:23  error  'console' is not defined    no-undef
  11:1   error  'beforeAll' is not defined  no-undef
  12:3   error  'console' is not defined    no-undef
  22:24  error  'console' is not defined    no-undef
  26:1   error  'afterAll' is not defined   no-undef
  27:3   error  'console' is not defined    no-undef

✖ 7 problems (7 errors, 0 warnings)

iamanishx commented:
> Can you add few tests to validate calculatehash function?

yes sure but the eslint ain't working and that may result failing ci tests what should i do ?

iamanishx commented:
@varshith257 is it good now ? i actually failed to push this file as my eslint was failing so now i have updated the `filehash.spec.ts`
and this PR related to this one 
https://github.com/PalisadoesFoundation/talawa-api/pull/3355
please review and let me know if i need to configure something 


here the tests will fail as i have updated the mutation according to api so @adithyanotfound  will fix the failing tests .
thank you.

Cioppolo14 commented:
@iamanishx The failed tests need to pass in order for us to move to the review stage. Please fix the failed test application check.

adithyanotfound commented:
@iamanishx Can you please replace the `mockUploadFileToMinio` with the following in line 6380 of `CreateGroupChat.spec.tsx`.

```
const mockUploadFileToMinio = vi.fn().mockResolvedValue({
    objectName: 'https://minio.example.com/test-image.jpg',
    fileHash: 'dummyFileHash',
});
```

This will fix the failing test and we can avoid reverting the PR. Thanks.

iamanishx commented:
> @iamanishx Can you please replace the `mockUploadFileToMinio` with the following in line 6380 of `CreateGroupChat.spec.tsx`.
> 
> 
> 
> ```
> 
> const mockUploadFileToMinio = vi.fn().mockResolvedValue({
> 
>     objectName: 'https://minio.example.com/test-image.jpg',
> 
>     fileHash: 'dummyFileHash',
> 
> });
> 
> ```
> 
> 
> 
> This will fix the failing test and we can avoid reverting the PR. Thanks.

Yeah sure thanks

iamanishx commented:
@Cioppolo14 In modern browsers, the File API—including arrayBuffer()—is fully implemented and works with the Web Crypto API to produce the correct hash. The test failures occur because our test environment (Node.js ) lacks a complete implementation of the File API, causing file.arrayBuffer() to be undefined. 
its the reason the patches are low .

iamanishx commented:
@palisadoes @varshith257 @noman2002  please review 
https://github.com/PalisadoesFoundation/talawa-admin/pull/3825
https://github.com/PalisadoesFoundation/talawa-admin/pull/3839
these PRs needs to configured according to this PR .

palisadoes commented:
Please get the code coverage of the patch closer to 100%. We cannot afford for code quality to decline

iamanishx commented:
> Please get the code coverage of the patch closer to 100%. We cannot afford for code quality to decline



> Please get the code coverage of the patch closer to 100%. We cannot afford for code quality to decline

again trying...

iamanishx commented:
@palisadoes is it okaayyyy now . please review

duplixx reviewed (approved):
LGTM
